### PR TITLE
Create Multimodel Build CI Testing Image

### LINKED ISSUE #9: Software installation in current images can only be done easily using Spack
Currently the packages get installed in directories whose name includes hashes. That means that the only practical way to use them is through Spack, i.e., one needs to do `spack load ...` before trying to use the packages in a build.

It would be nice to support other ways of building the software, for example, by using environment modules. Spack does install the module files, but they can't be used because neither lmod nor tcl modules are installed in the image.

Another way would be to install the packages as part of a Spack environment. This would provide a "view"  where all the packages are easily accessible. Better yet, check the Dockerfiles that Spack creates to automatically install an environment. It does it in a nice way that makes it very easy to use the packages and, at the same time, minimizes considerably the size of the docker images.

### PR DESCRIPTION
**Pull request by CodeGat**
After the discussions in https://forum.access-hive.org.au/t/ci-cd-planning-discussion/1035/3 we note that a multimodel, `spack env`-based Dockerfile is the way to go for future CI stuff. 
This will lead to a fair bit of refactoring and removal of much of the util folder. 

Should close #9 !

### CODE REVIEW

File: .github/workflows/build-and-push-image-build.yml
Comment (aidanheerdegen) [inline .github/workflows/build-and-push-image-build.yml:40]: Do we need an `if` at all if we're always running it?

File: .github/workflows/build-and-push-image-build.yml
Comment (aidanheerdegen) [inline .github/workflows/build-and-push-image-build.yml:54]: At the risk of adding more back into the container name, if we're going for monolithic "per-model" containers the model name should figure somewhere I think. Probably in the name, in which case does it also have to be an argument to the workflow under `jobs:`?

The ideal should be that we can specify a different model, e.g. access-om3, and all of this would work for that monolithic build  as well.

Don't rule out the possibility that I've misunderstood how this works and this is not required.

File: containers/Dockerfile.build
Comment (aidanheerdegen) [inline containers/Dockerfile.build:4]: So *if* what I was saying above about specifying the model above is correct, here some of the dependency mapping type stuff might still be useful. 
At the very least a `yaml` file something like
```yaml
access-om2: mom5 cice5 oasis3-mct libaccessom2
```

Could be defined and then used to populated the package names here? An alternative would be to have multiple `Dockerfile.build` files named for each model, e.g. `Dockerfile.access-om2.build`

File: .github/workflows/build-package.yml
Comment (aidanheerdegen) [inline .github/workflows/build-package.yml:64]: We're uploading logs of failed builds for debugging and so they don't get deleted when workflow logs are cleaned?

File: .github/workflows/build-package.yml
Comment (aidanheerdegen) [inline .github/workflows/build-package.yml:50]: We don't want to generate lockfile regardless of installation success or failure? I don't exactly know what I want to do with them though. :)

**Review by aidanheerdegen — APPROVED** [review]
Assuming what I said wasn't a misunderstanding I think it should be addressed before merging. 

Otherwise merge away!

File: .github/workflows/build-and-push-image-build.yml
Comment (aidanheerdegen) [reply .github/workflows/build-and-push-image-build.yml:40]: Answered my own question: this is required if a previous step might error and you still want this to run

see https://stackoverflow.com/a/58859404

Comment (CodeGat) [reply .github/workflows/build-and-push-image-build.yml:40]: Yeah this is for the case where we skip the above base-spack-build-and-push-image (because that base-spack image already exists). We still need it to have pushed the image before we do any of the other dependency image stuff, though.

File: .github/workflows/build-and-push-image-build.yml
Comment (CodeGat) [reply .github/workflows/build-and-push-image-build.yml:54]: No, you're correct there. We'll need to specify the overall model name (access-om2, access-om3) in the build. Will update now

File: .github/workflows/build-package.yml
Comment (CodeGat) [reply .github/workflows/build-package.yml:50]: Might be worth doing them on both conditions, actually. Will update the repo.

File: .github/workflows/build-package.yml
Comment (CodeGat) [reply .github/workflows/build-package.yml:64]: Not the logs, just the lockfiles. I think it'd be useful to see what the difference is between the lockfile that we have, and the lockfile that it regenerates to install it correctly, so we could update dependencies

File: containers/Dockerfile.build
Comment (CodeGat) [reply containers/Dockerfile.build:4]: Dependency mapping still has a home, it seems! I'll update the build.

File: .github/workflows/build-package.yml
Comment (aidanheerdegen) [reply .github/workflows/build-package.yml:64]: Agreed. I don't know if "logs" was a typo on my part or auto-correct, but I did understand they were lockfiles.

File: containers/Dockerfile.build
Comment (aidanheerdegen) [reply containers/Dockerfile.build:4]: > Dependency mapping still has a home, it seems! I'll update the build.

Or have multiple Dockerfiles. It's not a terrible option.

File: .github/workflows/build-and-push-image-build.yml
Comment (CodeGat) [reply .github/workflows/build-and-push-image-build.yml]: Updated!

File: containers/Dockerfile.build
Comment (CodeGat) [reply containers/Dockerfile.build]: Added multiple dockerfiles :)

File: .github/workflows/build-package.yml
Comment (CodeGat) [reply .github/workflows/build-package.yml]: Added on both conditions

File: containers/Dockerfile.access-om3
Comment (aidanheerdegen) [inline containers/Dockerfile.access-om3:4]: To be consistent with `access-om2` we should use the model components as defined in here

https://github.com/COSIMA/access-om3/tree/master

Namely: CDEPS, CICE6, CMEPS, MOM6, WW3

But we don't have package files for any of those so we can't put them in there. So `access-om3` is a decent place-holder as it just defines a few dependencies.

File: containers/compilers.yaml
Comment (aidanheerdegen) [inline containers/compilers.yaml]: I definitely want the next PR to add a compiler to this and check it works ;)

**Review by aidanheerdegen — APPROVED** [review]
Looks good, just a couple of vague comments.

**Comment by aidanheerdegen** [comment]
Feel free to resolve those conversations that are blocking the merge (once you're read them of course ;) )

File: containers/compilers.yaml
Comment (CodeGat) [reply containers/compilers.yaml]: Will give that a go!

File: containers/Dockerfile.access-om3
Comment (CodeGat) [reply containers/Dockerfile.access-om3:4]: Noted, will keep in mind when we add package files

**Pull request merged**
